common/src/hashing_domain.rs

```diff
@@ -0,0 +1,172 @@
+// Copyright 2019. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use sha2::Digest;
+use tari_common_types::types::{GenericDomainHasher, MacDomainHasher};
+use tari_crypto::hashing::{DomainSeparatedHash, LengthExtensionAttackResistant, Mac};
+use thiserror::Error;
+
+pub struct HashingDomain {
+    domain_label: &'static str,
+}
+
+/// Error type for the pipeline.
+#[derive(Debug, Error)]
+pub enum HashingDomainError {
+    #[error("Two slices have different lengths")]
+    CopyFromSlice,
+}
+
+impl HashingDomain {
+    /// A new constant hashing domain with domain label
+    pub const fn new(domain_label: &'static str) -> Self {
+        Self { domain_label }
+    }
+
+    /// A new generic domain separated hasher for the chosen hashing domain
+    pub fn hasher<D: Digest>(&self) -> GenericDomainHasher<D> {
+        GenericDomainHasher::new(self.domain_label)
+    }
+
+    /// Convenience function to compute hash of the data. It will handle hasher creation, data feeding and finalization.
+    pub fn digest<D: Digest>(&self, data: &[u8]) -> DomainSeparatedHash {
+        self.hasher::<D>().chain(data).finalize()
+    }
+
+    /// Helper function to return a fixed 32 size hashed bytes array
+    pub fn hash_to_bytes_32(hash: &DomainSeparatedHash) -> Result<[u8; 32], HashingDomainError> {
```

Comment:
How about yo add a trait for that and implement for the `DomainSeparatedHash`? [[demo](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a82d0ad66d95d484a7d809363912a9aa)]

```rust
pub trait HashToBytes<const I: usize>: AsRef<[u8]> {
    fn hash_to_bytes(&self) -> Result<[u8; I], HashingDomainError> {
        let hash_vec = self.as_ref();
        if hash_vec.is_empty() || hash_vec.len() < I {
            return Err(HashingDomainError::CopyFromSlice);
        }
        let mut buffer: [u8; I] = [0; I];
        buffer.copy_from_slice(&hash_vec[..I]);
        Ok(buffer)
    }
}

// It has to be implemented for the `DomainSeparatedHash` instead
impl<const I: usize> HashToBytes<I> for Vec<u8> {}

pub fn somewhere() -> Result<(), Error> {
    let data = vec![0u8,1,2,3,4,5,6,7,8];
    let bytes: [u8; 8] = data.hash_to_bytes()?;
    let bytes: [u8; 4] = data.hash_to_bytes()?;

    // Error is thrown here (not enough bytes the slice has)
    let bytes: [u8; 16] = data.hash_to_bytes()?;
    Ok(())
}
```